### ISSUE TITLE
[regression] .mean() for half tensors no longer gives correct result even if result is representable in fp16

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#12115]
```
import torch
x = torch.ones(65520, device='cuda', dtype=torch.float16)
x.mean()
```
used to produce `tensor(1., dtype=torch.float16, device='cuda:0')` around the end of May, now it is 
`tensor(nan, device='cuda:0', dtype=torch.float16)`
Two reasons: mean was moved to ATen and thus no longer uses fused THC kernels for this
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L96-L109 (that alone would lead to result being `inf`), and long `numel` is apparently being cast to half somewhere in TensorIterator `div_`, thus leading to `nan` answer.
Our fault for not adding the tests at the time @mruberry fixed it.

**Comment by mruberry** [comment]
(our fault = my fault)

There is an fp16 mean test but it doesn't catch this failure.

**Comment by yaysummeriscoming** [comment]
Hi, as well as the overflow issue there also seems to be an issue with accuracy:

Compute mean in 2 stages here so that we don't overflow fp16
def _fp16_safe_mean(tensor):
    tensor = tensor.mean(dim=0)
    tensor = tensor.mean(dim=0)
    tensor = tensor.mean()
    return tensor

x = torch.FloatTensor(np.random.rand(16, 28, 28, 8))
x = x.abs()
x = x.type(torch.float16)
x = x.cuda()
print('FP16 safe mean is: %f' % _fp16_safe_mean(x))
print('Normal mean is: %f' % x.mean())
print('FP16 sum is: %f' % (x.sum()))

x = x.type(torch.float32)
print('Float32 reference mean is: %f' % x.mean())

FP16 safe mean is: 0.500488
Normal mean is: 0.000000
FP16 sum is: 50240.000000
Float32 reference mean is: 0.500724

So it seems that the summing is accurate, but somewhere in the division stage of the mean function something is going wrong?

**Comment by ngimel** [comment]
Yep, number of your tensor elements is more than fp16 max, and it overflows when it is converted to half to calculate mean. Then 1/num_elements becomes 0.

**Comment by yaysummeriscoming** [comment]
Sorry if I should be opening a new issue for this, but I did some further testing of mean() vs F.avg_pool2d().  I'm interested because mean() seems to be much faster (#14615).  I find that using mean() instead of F.avg_pool2d() results in a very slight accuracy drop during fp16 training:

x = torch.FloatTensor(np.random.rand(64, 64, 28, 28))
x = x.type(torch.float16)
x = x.cuda()
x_reshaped = x.view(*(x.shape[:-2]), -1)

fp32_mean = x_reshaped.type(torch.float32).mean(-1).half()
normal_mean = x_reshaped.mean(-1)
alt_mean = (x_reshaped.sum(-1).float() / (28 * 28)).half()
av_pool = F.avg_pool2d(x, kernel_size=28).squeeze()
normal_mean_diff = ()

def _diff(x1, x2):
    return (x1 - x2).abs().sum()

print('Difference between fp32 mean and F.avg_pool2d is: %f' % _diff(fp32_mean, av_pool))
print('Difference between fp32 mean and normal mean is: %f' % _diff(fp32_mean, normal_mean))
print('Difference between fp32 mean and mean divided in fp32 precision is: %f' % _diff(fp32_mean, alt_mean))

Output:
Difference between fp32 mean and F.avg_pool2d is: 0.000000
Difference between fp32 mean and normal mean is: 0.695312
Difference between fp32 mean and mean divided in fp32 precision is: 0.329834

It looks like F.avg_pool2d() converts internally to fp32?

**Comment by ngimel** [comment]
@yaysummeriscoming you are right, what you are seeing is part of the same regression. Before it, mean was calculated in the following way
1) fp16 data was converted to fp32 on read
2) sum was computed in fp32
3) in the same kernel, with sum remaining in fp32 and number of elements in fp32, mean was calculated
4) final value was converted to fp16 and written out. 
Looks like avg_pool2d still retains this behavior, but mean regressed and does the following https://github.com/pytorch/pytorch/blob/35c8f93fd238d42aaea8fd6e730c3da9e18257cc/aten/src/ATen/native/ReduceOps.cpp#L178-L191
1) fp16 data is still converted to fp32 on read
2) sum is computed in fp32, but then is converted to fp16 and written out
3) 1/num_elements is converted to fp16, and then sum is multiplied by it (that's how `div_` works).
2) and 3) lead to accuracy regressions. When you divide in fp32, you are getting rid of errors caused by 3). For now, there's no good workaround (converting data to float and computing float mean would be slower than calling avg_pool2d). @jjsjann123 is working on restoring the previous behavior, but it will be some time. If you want, you can coordinate with him and take a stab at it :-)

**Comment by yaysummeriscoming** [comment]
@ngimel, thanks for the detailed explanation & pointing me to the relevant source code.  I jumped ship from tensorflow a few months ago, so I'm not very familiar with it yet.  I'm a bit hesitant touching source code until I get a bit more familiar, but perhaps I could be of assistance with docs & testing.

**Comment by jjsjann123** [comment]
Hey no worries. I'm pushing PRs to clean up these reduction related issues.
Really appreciate your report on the problem and hope to see you active in this community :)

**Comment by yaysummeriscoming** [comment]
Thanks, you most certainly will!  I figure I'm going to start with some cpp extensions when I get some time.  Still coming to terms with how much easier to use pytorch is than tensorflow!

**Comment by vishwakftw** [comment]
I think this can be closed since #14878 has been merged.
